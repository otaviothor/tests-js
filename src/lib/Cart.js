export default class Cart {
  items = [];

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  add(item) {
    this.items.push(item);
  }

  remove(product) {}

  summary() {}

  checkout() {}
}

const cart = new Cart();

const product = {
  title: '',
  price: '',
};

const item = {
  quantity: 2,
  product,
};

cart.add(item);
cart.remove(product);
cart.getTotal();
cart.summary();
cart.checkout();
