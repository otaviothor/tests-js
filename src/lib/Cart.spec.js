import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product;
  let product2;

  beforeEach(() => {
    cart = new Cart();
    product = {
      title: 'Adidas Running - Men',
      price: 35388, // 353.88 | R$ 353,88
    };
    product2 = {
      title: 'Adidas Running - Women',
      price: 41872,
    };
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance ', () => {
      expect(cart.getTotal()).toEqual(0);
    });

    it('should multiply quantity and price and recieve the total amount', () => {
      const item = {
        product,
        quantity: 2, // 70776
      };

      cart.add(item);

      expect(cart.getTotal()).toEqual(70776);
    });

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal()).toEqual(35388);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal()).toEqual(41872);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
      expect(cart.checkout()).toMatchInlineSnapshot(`
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas Running - Men",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas Running - Women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`);
    });
  });
});
