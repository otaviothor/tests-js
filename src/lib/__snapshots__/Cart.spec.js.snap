// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas Running - Men",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas Running - Women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
